### PR TITLE
Fix not all traits respecting multiple instances of 'Production'

### PR DESCRIPTION
**Pull request by abcdefg30**
~~I'll update `ClonesProducedUnits` in a follow-up PR.~~

> Note for clarity: Currently the last Harkonnen missions crash (and that's probably just one case; #14796 "fixed" this for the upcoming playtest.) We want to sort all those issues with multiple instances of `Produces` out for `Next+1`. (I.e. replace all occurrences of `Trait<Produces>` and `TraitInfo<ProducesInfo>`)

### CODE REVIEW

**Comment by abcdefg30** [comment]
We decided to revert the commit introducing the issue in the meantime: #14796

**Comment by abcdefg30** [comment]
Note for clarity: Currently the last Harkonnen missions crash (and that's probably just one case; #14796 "fixed" this for the upcoming playtest.) We want to sort all those issues with multiple instances of `Produces` out for `Next+1`. (I.e. replace all occurrences of `Trait<Produces>` and `TraitInfo<ProducesInfo>`)

**Comment by GraionDilach** [comment]
Why is 9e2b55a needed?

**Comment by pchote** [comment]
It was a test case commit that was not removed before merging.  It makes the build-everything cheat more difficult to use because it shows as an exact duplicate of the Harkonnen conyard in the sidebar.  It doesn’t make sense / offer any real benefit to allow the special palace to be built using cheats in MP, especially since we do not do the same for the other special factions.

**Comment by pchote** [comment]
Do you have a lua script to exercise the testcase? The lua integration is the main thing that needs testing here.

**Comment by abcdefg30** [comment]
Updated.

File: OpenRA.Mods.Common/Traits/Render/ProductionBar.cs
Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/Render/ProductionBar.cs]: Why do we need this variable?

File: OpenRA.Mods.Common/Traits/Render/ProductionBar.cs
Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/Render/ProductionBar.cs]: Related to my other nit: wouldn't it make more sense to just FieldLoader.Require the queue type explicitly?

File: OpenRA.Mods.Common/Traits/Render/ProductionBar.cs
Comment (abcdefg30) [reply OpenRA.Mods.Common/Traits/Render/ProductionBar.cs]: How else do you suggest I handle the case when `queue == null` without reassigning the `queue` variable? (We could e.g. move the `value` assignment in if statements, I thought this is the cleaner approach though.)

File: OpenRA.Mods.Common/Traits/Render/ProductionBar.cs
Comment (abcdefg30) [reply OpenRA.Mods.Common/Traits/Render/ProductionBar.cs]: Maybe; that sounds like a good idea to simplify the code here. Will have a look.

**Comment by abcdefg30** [comment]
Updated.

File: OpenRA.Mods.Common/Traits/Render/ProductionBar.cs
Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/Render/ProductionBar.cs]: Why is this removed? This breaks captured ClassicProdQueues I think.

File: OpenRA.Mods.Common/Traits/Render/ProductionBar.cs
Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/Render/ProductionBar.cs]: You should also keep this check around, incase the modder typo.

File: OpenRA.Mods.Common/UpdateRules/Rules/RequireProductionType.cs
Comment (pchote) [inline OpenRA.Mods.Common/UpdateRules/Rules/RequireProductionType.cs]: "The value be"

File: OpenRA.Mods.Common/Widgets/Logic/Ingame/Hotkeys/CycleProductionActorsHotkeyLogic.cs
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/Ingame/Hotkeys/CycleProductionActorsHotkeyLogic.cs]: This will throw an exception if all traits are disabled

**Comment by pchote** [comment]
While this would be good to have, I don't think it is critical so bumping off the milestone.   If you can do the fixup/rebase in time then we should still try and get this in.

**Comment by reaperrr** [comment]
I'm not entirely sure what bleed looks like in that regard (since this PR hasn't been rebased in ages), but we'll want https://github.com/OpenRA/OpenRA/pull/14795/commits/97cb4807e347c8311d03e6985e3c505da1cef40e if it's not on bleed yet (iirc it was picked to prep-1801 directly).

File: OpenRA.Mods.Common/Traits/Render/ProductionBar.cs
Comment (pchote) [reply OpenRA.Mods.Common/Traits/Render/ProductionBar.cs]: We (should) have lint tests for that!

**Comment by pchote** [comment]
Added back to the milestone because of the D2K mission crashes. But IMO it would be safer to cherry pick the fix from the last prep branch instead.

**Comment by abcdefg30** [comment]
Resolved the requests and fixed primary building.

**Comment by GraionDilach** [comment]
> it would be safer to cherry pick the fix from the last prep branch instead.

Agreeing as the solution for the release tbh.

**Comment by abcdefg30** [comment]
Rebased.

**Comment by pchote** [comment]
Needs a rebase, and the update rule should move to the new directory.

**Comment by abcdefg30** [comment]
Rebased.

**Comment by pchote** [comment]
I rebased this (on top of #15743 to get the worst of the rebase conflicts out of the way at once), reordered a few things to simplify reviewing, and fixed a few bugs that I noticed.

The code looks mostly ok, but needs testing before giving a final review.

**Review by pchote — APPROVED** [review]
Rebased again, split out specific commits for the different changes, and tested as well as I can.

LGTM, but would like a second review to cover my changes here before we merge.

**Comment by abcdefg30** [comment]
Your changes look good to me. 👍 / ✅

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/d728a12c8f2fdd422fa9e521bbe00595f8caaad1...6ffc68eebdd01a641bb5272b750de2854eba7a90)

**Pull request merged**
